File: 2022/09/PHP/main.php
<?php

class Knot {
	private $row;
	private $col;
	private $history;

	public function __construct( $row = 0, $col = 0 ) {
		$this->row = $row;
		$this->col = $col;
		$this->history = [];
		$this->track_history();
	}

	private function track_history () {
		$key = implode( ',', [ $this->row, $this->col ] );
		if ( empty( $this->history[$key] ) )
			$this->history[$key] = 0;
		++$this->history[$key];
	}

	public function history_size () {
		return count( $this->history );
	}

	public function move ( $direction ) {
		str_contains( $direction, 'U' ) and --$this->row;
		str_contains( $direction, 'D' ) and ++$this->row;
		str_contains( $direction, 'L' ) and --$this->col;
		str_contains( $direction, 'R' ) and ++$this->col;
		$this->track_history();
	}

	public function follow ( $other ) {
		if ( abs( $this->row - $other->row ) <= 1 and abs( $this->col - $other->col ) <= 1 ) {
			return;
		}
		$direction = '';
		$this->row > $other->row and $direction .= 'U';
		$this->row < $other->row and $direction .= 'D';
		$this->col > $other->col and $direction .= 'L';
		$this->col < $other->col and $direction .= 'R';
		$this->move( $direction );
	}
}

function main ( $filename, $knot_count = 2, $desc = "ANSWER" ) {
	$knot_count > 1 or die();
	for ( $k = 0; $k < $knot_count; ++$k ) {
		$knots[] = new Knot( 0, 0 );
	}
	$lines = file( $filename );
	foreach ( $lines as $line ) {
		list ( $direction, $move_count ) = explode( ' ', trim( $line ) );
		while ( $move_count-->0 ) {
			foreach ( $knots as $ix => $knot ) {
				if ( $ix == 0 ) {
					$knot->move( $direction );
					continue;
				}
				$knot->follow( $knots[$ix - 1] );
			}
		}
	}
	echo "$desc: " . $knots[$knot_count-1]->history_size() . "\n";
}

main( getenv('ADVENT_INPUT'), 2, "PART1" );
main( getenv('ADVENT_INPUT'), 10, "PART2" );

File: 2022/09/Rust/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};

static INPUT: &str = env!("ADVENT_INPUT");

type MoveHistory = HashMap<(i32, i32), usize>;

#[derive(Clone)]
struct Knot {
    row: i32,
    col: i32,
    history: MoveHistory,
}

impl Knot {
    pub fn new(row: i32, col: i32) -> Self {
        let history: MoveHistory = HashMap::new();
        let mut new = Self { row, col, history };
        new.track_history();
        new
    }

    pub fn track_history(&mut self) {
        let key = (self.row, self.col);
        *self.history.entry(key).or_insert(0) += 1;
    }

    pub fn history_size(&self) -> usize {
        self.history.len()
    }

    pub fn budge(&mut self, direction: &str) {
        if direction.contains("U") {
            self.row -= 1;
        }
        if direction.contains("D") {
            self.row += 1;
        }
        if direction.contains("L") {
            self.col -= 1;
        }
        if direction.contains("R") {
            self.col += 1;
        }
        self.track_history();
    }

    pub fn follow(&mut self, other: &Self) {
        if (self.row - other.row).abs() <= 1 && (self.col - other.col).abs() <= 1 {
            return;
        }
        let mut direction = String::new();
        if self.row > other.row {
            direction.push_str("U");
        }
        if self.row < other.row {
            direction.push_str("D");
        }
        if self.col > other.col {
            direction.push_str("L");
        }
        if self.col < other.col {
            direction.push_str("R");
        }
        self.budge(&direction);
    }
}

pub fn solve(filename: &str, knot_count: usize, desc: &str) {
    if knot_count < 1 {
        panic!("knot_count too low");
    }
    let mut knots: Vec<Knot> = (1..=knot_count).map(|_| Knot::new(0, 0)).collect();

    let file = File::open(filename).unwrap();
    let io = BufReader::new(file);
    for line in io.lines() {
        let text = line.unwrap();
        let parts: Vec<&str> = text.split(" ").collect();
        let direction = parts[0];
        let move_count = parts[1].parse().unwrap();
        for _ in 0..move_count {
            knots[0].budge(direction);
            for j in 1..knot_count {
                let x = knots[j - 1].clone();
                knots[j].follow(&x);
            }
        }
    }
    println!("{}: {}", desc, knots[knot_count - 1].history_size());
}

pub fn main() {
    solve(INPUT, 2, "PART1");
    solve(INPUT, 10, "PART2");
}
